File: Tests/kernels_loop.c
#include "acc_testsuite.h"

int test(){
    int err = 0;
    srand(time(NULL));
    
    real_t * a = (real_t *)malloc(1024 * sizeof(real_t));
    real_t * b = (real_t *)malloc(1024 * sizeof(real_t));

    for (int x = 0; x < 1024; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0.0;
    }

    #pragma acc kernels copy(a[0:1024]) copyout(b[0:1024])
    {
        #pragma acc loop
        for (int _0 = 0; _0 < 2; ++_0){
            #pragma acc loop
            for (int _1 = 0; _1 < 2; ++_1){
                #pragma acc loop
                for (int _2 = 0; _2 < 2; ++_2){
                    #pragma acc loop
                    for (int _3 = 0; _3 < 2; ++_3){
                        #pragma acc loop
                        for (int _4 = 0; _4 < 2; ++_4){
                            #pragma acc loop
                            for (int _5 = 0; _5 < 2; ++_5){
                                #pragma acc loop
                                for (int _6 = 0; _6 < 2; ++_6){
                                    #pragma acc loop
                                    for (int _7 = 0; _7 < 2; ++_7){
                                        #pragma acc loop
                                        for (int _8 = 0; _8 < 2; ++_8){
                                            #pragma acc loop
                                            for (int _9 = 0; _9 < 2; ++_9){
                                                b[_0 * 512 + _1 * 256 + _2 * 128 + _3 * 64 + _4 * 32 + _5 * 16 + _6 * 8 + _7 * 4 + _8 * 2 + _9] = a[_0 * 512 + _1 * 256 + _2 * 128 + _3 * 64 + _4 * 32 + _5 * 16 + _6 * 8 + _7 * 4 + _8 * 2 + _9];
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    
    for (int x = 0; x < 1024; ++x){
        if (fabs(a[x] - b[x]) > PRECISION){
            err += 1;
        }
    }
    
    free(a);
    free(b);
    return err;
}


int main()
{
  int i;			/* Loop index */
  int result;		/* return value of the program */
  int failed=0; 		/* Number of failed tests */
  int success=0;		/* number of succeeded tests */
  static FILE * logFile;	/* pointer onto the logfile */
  static const char * logFileName = "test_acc_lib_acc_wait.log";	/* name of the logfile */


  /* Open a new Logfile or overwrite the existing one. */
  logFile = fopen(logFileName,"w+");

  printf("######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  printf("## Repetitions: %3d                       ####\n",REPETITIONS);
  printf("## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  printf("##############################################\n");
  printf("Testing test_acc_lib_acc_wait\n\n");

  fprintf(logFile,"######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  fprintf(logFile,"## Repetitions: %3d                       ####\n",REPETITIONS);
  fprintf(logFile,"## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  fprintf(logFile,"##############################################\n");
  fprintf(logFile,"Testing test_acc_lib_acc_wait\n\n");

  for ( i = 0; i < REPETITIONS; i++ ) {
    fprintf (logFile, "\n\n%d. run of test_acc_lib_acc_wait out of %d\n\n",i+1,REPETITIONS);
    if (test() == 0) {
      fprintf(logFile,"Test successful.\n");
      success++;
    } else {
      fprintf(logFile,"Error: Test failed.\n");
      printf("Error: Test failed.\n");
      failed++;
    }
  }

  if(failed==0) {
    fprintf(logFile,"\nDirective worked without errors.\n");
    printf("Directive worked without errors.\n");
    result=0;
  } else {
    fprintf(logFile,"\nDirective failed the test %i times out of %i. %i were successful\n",failed,REPETITIONS,success);
    printf("Directive failed the test %i times out of %i.\n%i test(s) were successful\n",failed,REPETITIONS,success);
    result = (int) (((double) failed / (double) REPETITIONS ) * 100 );
  }
  printf ("Result: %i\n", result);
  return result;
}

File: Tests/kernels_loop.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
         INTEGER :: x, v_0, v_1, v_2, v_3, v_4, v_5, v_6, v_7, v_8, &
            v_9 !Iterators
        REAL(8),DIMENSION(1024):: a, b !Data
        INTEGER :: errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        b = 0

        !$acc data copyin(a(1:1024)) copyout(b(1:1024))
          !$acc kernels
            !$acc loop
            DO v_0 = 0, 1
              !$acc loop
              DO v_1 = 0, 1
                !$acc loop
                DO v_2 = 0, 1
                  !$acc loop
                  DO v_3 = 0, 1
                    !$acc loop
                    DO v_4 = 0, 1
                      !$acc loop
                      DO v_5 = 0, 1
                        !$acc loop
                        DO v_6 = 0, 1
                          !$acc loop
                          DO v_7 = 0, 1
                            !$acc loop
                            DO v_8 = 0, 1
                              !$acc loop
                              DO v_9 = 1, 2
                                 b(v_0 * 512 + v_1 * 256 + v_2 * 128 &
                                    + v_3 * 64 + v_4 * 32 + v_5 * 16 &
                                    + v_6 * 8 + v_7 * 4 + v_8 * 2 + &
                                    v_9) = a(v_0 * 512 + v_1 * 256 + &
                                    v_2 * 128 + v_3 * 64 + v_4 * 32 + &
                                    v_5 * 16 + v_6 * 8 + v_7 * 4 + &
                                    v_8 * 2 + v_9)
                              END DO
                            END DO
                          END DO
                        END DO
                      END DO
                    END DO
                  END DO
                END DO
              END DO
            END DO
          !$acc end kernels
        !$acc end data

        DO x = 1, 1024
          IF (abs(a(x) - b(x)) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of kernels_loop"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of kernels_loop"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM


